refactor: enforce usage of `Pokemon.isOfType` over `Pokemon.getTypes`; make default values consistent

Bertie690 opened the pull request:
## What are the changes the user will see?
A few moves and abilities will now have correct interactions when Stellar Terastallized.
- Flower Veil
- Flower Shield
- Reflect Type

A few other moves were changed to properly fail if the user is Terastallized:
- Conversion 2
- Conversion (also fails if user is same type as its first move, for consistency)

Finally, Forest's Curse/Trick-or-Treat will be automatically removed when the user Terastallizes (tested on showdown (<https://replay.pokemonshowdown.com/gen9ou-2530172961-8xgqrtwj8acyevf4srfv4m8vyohs7e9pw>). These effects were previously ignored if the user was Terastallized, but this saves a bit of space in save data.

## Why am I making these changes?
I wanted to review the default parameter state of `Pokemon#getTypes`.
Then I realized its defaults were different from `Pokemon#isOfType` (which is outright confusing), alongside being a 71 cognitive complexity MESS.

> [!NOTE]
> I am keeping this PR small and self-contained as I want to make `forDefend` default to `true` later on as well (and perhaps even remove it due to being useful in exactly one place) 

## What are the changes from a developer perspective?
Cleaned up implementation of `Pokemon.getTypes`, and  changed its `useTeraType` parameter to `true` (matching `isOfType`'s default behavior).

Fixed required spots where it was being passed without any arguments (all that remain now either don't run if the user is Terastallized or explicitly disregard it).

## Screenshots/Videos
N/A? Annoyingly hard to test

## How to test the changes?
`pnpm test flower-shield.test.ts` (a few new tests, including a Tera Stellar case)

## Checklist
- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes manually?
- [x] Are all unit tests still passing? (`pnpm test:silent`)
  - [x] Have I created new automated tests (`pnpm test:create`) or updated existing tests related to the PR's changes?

Bertie690 committed:
fixed several instances of not checking tera types

Bertie690 committed:
[Refactor] Enforced usage of `isOfType` over `getTypes`; made default values consistent

Bertie690 committed:
Enforced matcher usage again

Bertie690 committed:
Fixed trnasform tests

Bertie690 committed:
more test fixes

Bertie690 committed:
Added fail condition to Conversion

Bertie690 committed:
fixed flower gift tests

Bertie690 committed:
fixed syntax errors

Bertie690 committed:
Update to-have-types.ts

Bertie690 committed:
Update to-have-types.ts

Bertie690 committed:
Fix even more assorted bugs with typing mismatches

Bertie690 committed:
comment fix

Bertie690 committed:
Merge branch 'beta' into for-defend

Bertie690 committed:
Fix import issue

Bertie690 committed:
Tix flower-shield.test.ts

Bertie690 committed:
Merge branch 'beta' into for-defend

Bertie690 committed:
Update game ingo get types call

Bertie690 committed:
Merge remote-tracking branch 'upstream/beta' into for-defend

Bertie690 committed:
ran boime

DayKev committed:
Merge branch 'beta' into for-defend

Bertie690 committed:
Merge branch 'beta' into for-defend

Bertie690 committed:
Merge branch 'beta' into for-defend

Bertie690 committed:
Merge remote-tracking branch 'upstream/beta' into for-defend

Bertie690 committed:
1 fix stuff

Wlowscha commented:
I've been digging around a bit and I have some notes about `getTypes`:

The `forDefend` parameter is only used in relation to tera types. It has no effect otherwise, so for any calculations where `includeTeraType` is `false`, the value of `forDefend` is irrelevant. In fact, there is only a single callsite where `getTypes(false, true,` is used.

When `includeTeraType` is `true`, we have this logic
```if (forDefend) {
  return [teraType];
}
types.push(teraType);
```
However, this is arguably pointless since in the second case, no other types are added afterwards; the original types are on a different conditional branch, and extra types (e.g. from Trick or Treat) should be irrelevant in the case of tera. This means that in practice, the `forDefend` parameter is only relevant in the case of TeraStellar, to determine whether the Stellar type OR the original types should be returned.

In most cases we call `getTypes(true, true,` which returns the original types.

Conversely, `getTypes(true, false,` only appears at three callsites:
1) In `Pokemon.getMatchupScore`, where the score is influenced by the effectiveness of the enemy's types, so we do want `getTypes` to return Stellar;
2) In two methods of the `BattleInfo` class, where again it makes sense to have Stellar.

There are a few more cases however which call `getTypes(true)` (using `false` as the default):
1) When determining the effect of Curse (in `CurseAttr`)
2) In the `RemoveTypeAttr` class, but this looks like a mistake, because there is an explicit check for tera before that;
3) In `genNewPartyMemberSpecies`, where I am not sure how it would be relevant.

As far as I can see, `isOfType` is always called either with `true, true` or `false, false`.

Overall, this can lead to inconsistencies, see e.g. this comment in `move.ts`:
```
    new AttackMove(MoveId.BURN_UP, PokemonType.FIRE, MoveCategory.SPECIAL, 130, 100, 5, -1, 0, 7)
      // Pass `true` to `ForDefend` as it should fail if the user is terastallized to a type that is not FIRE
      .condition(user => user.isOfType(PokemonType.FIRE, true, true), 2)
```
Passing `true` to `forDefend` here means that the user is counted as fire type if it is currently Tera Stellar, which may not be intended.

--------------------

Based on this I would suggest the following changes, which may fit in this PR:
1) Clear up the `if (forDefend)` conditional in `getTypes`
2) Rename the `forDefend` parameter to something like `returnOriginalTypesIfStellar`, since that's the only case where it is relevant
3) Possibly, clean up the callsites indicated above if needed, although this might extend beyond the scope of this PR.

Otherwise, I will put this information in an issue and we can deal with it at a later time.

Bertie690 commented on src/field/pokemon.ts:
```diff
@@ -1935,114 +1938,45 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
+    includeTeraType = true,
     forDefend = false,
     ignoreOverride = false,
     useIllusion = false,
```
Correct.

I'm going to remove the `forDefend` parameter in a follow-up PR (for which I'll need to alter all the callsites anyways).

Bertie690 reviewed (commented):
> When determining the effect of Curse (in CurseAttr)

I kep that as is since one of my other PRs handling Curse explicitly fixes that

Bertie690 committed:
Apply the review comments

Bertie690 committed:
Merge remote-tracking branch 'upstream/beta' into for-defend

Bertie690 committed:
Fix merge conflict eating like 5 move attrs

Wlowscha committed:
Fixed fusion secondary typing

Bertie690 committed:
Merge remote-tracking branch 'upstream/beta' into for-defend

Bertie690 committed:
fix import order

Bertie690 commented:
@Wlowscha I have not forgotten about the shadow tag oversight, will checkout branch and fix tmr morning

Bertie690 committed:
Merge remote-tracking branch 'upstream/beta' into for-defend

Bertie690 committed:
fix shadow trap oversight

Bertie690 committed:
Merge branch 'beta' into for-defend

Bertie690 committed:
Merge remote-tracking branch 'upstream/beta' into for-defend

Bertie690 committed:
Merge branch 'beta' into for-defend

Bertie690 committed:
Fix comment borkage

DayKev committed:
Merge branch 'beta' into for-defend

DayKev committed:
Review changes

- Add/update some TSDocs/comments

- Fix type check in `mystery-encounter-requirements.ts`
to not falsely exclude Normal

- Add ability for `toHaveTypes` test matcher to check whether the
Pokemon has at least one of the expected types

- Add tests for Conversion and Conversion 2

DayKev commented on src/field/pokemon.ts:
```diff
@@ -3661,19 +3629,35 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
-    return Math.min(stabMultiplier.value, 2.25);
+    // TODO: Instead of ignoring terapagos' tera stellar boosts when calling this,
+    // we should avoid pushing its usages to the array altogether to reduce save data size
+    const canBoostStellar = !this.stellarTypesBoosted.includes(moveType) || this.hasSpecies(SpeciesId.TERAPAGOS);
```
Shouldn't this also be checking the attacker and not "`this`"?
You even made a comment on L3610: `// TODO: This uses nothing from this Pokemon AT ALL, so why is this called on the defender?`

Bertie690 replied:
Actually, I realize this is 100% correct - this method is called on `source` (attacker), not the defender (since it uses exactly 0 things from said defender).

I have updated the doc comment to mention this discrepancy for future notice.

DayKev replied:
Oh I see; that's confusing lol. Both methods should probably just be entirely removed from the `Pokemon` class.

DayKev committed:
Add test for Terastallization STAB boosts

Bertie690 committed:
Fix bug with tera stellar

DayKev committed:
Merge branch 'beta' into for-defend
